### PR TITLE
Enables deferred intent UPE for existing legacy UPE stores

### LINKED ISSUE #7220: Enable Deferred Intent UPE for legacy UPE users
### Description
Similar to https://github.com/Automattic/woocommerce-payments/issues/7003, but this time instead of newly onboarded merchants, legacy UPE merchants should have the deferred UPE feature flag enabled by default.

💡 The solution for this task **should be merged between Sep 19 and October 6,** so that it is released as part of WooPayments 6.6.0

### PR DESCRIPTION
**Pull request by timur27**
Closes https://github.com/Automattic/woocommerce-payments/issues/7220

> 💡 This PR needs to be reviewed & tested in pair with https://github.com/Automattic/woocommerce-payments-server/pull/3963

#### Changes proposed in this Pull Request
This PR accomplishes the goal of enabling deferred intent creation UPE for legacy UPE stores that use WooPayments of version `6.6.0` or higher. Previously, in `6.4.0`, we enabled deferred intent creation UPE for split UPE stores (https://github.com/Automattic/woocommerce-payments/pull/7058), and this PR was supposed to be pretty similar to the one for split UPE migration, but it eventually ended up being different. 

The main reason is that, unlike split UPE stores, legacy UPE stores migrated should be able to roll back to the legacy UPE if needed (by us, not by merchants). As part of the `legacy UPE` -> `deferred UPE` migration, the client-side feature flag is not updated from the UI, but instead, split UPE stores are treated as deferred UPE stores, which means `1 0` flag values for split UPE & deferred UPE accordingly. A similar implementation would be needed for the legacy UPE if we would rely only on the client-side flag. However, there's a small challenge.

This makes it difficult to roll back. Since the migrated store has feature flags set to `1 0` for legacy UPE & deferred UPE, what values should this pair have after rolling back to legacy UPE? Should we introduce `1 disabled` for this scenario? Maybe. https://github.com/Automattic/woocommerce-payments/pull/7278 tried to accomplish this, but the logic became pretty complex. This approach also requires us to reach out to merchants and ask them to perform specific actions to disable the flag. On top of that, deferred UPE enablement is coupled with the WooPayments plugin version and thus, any changes on this front would require stores to upgrade. 

Instead, we started to think about server-side feature flags and this PR is all about it. It is way easier to maintain the state of deferred UPE for particular stores. It's also easier to disable deferred UPE for everyone on the server side without releasing a new patch version and asking merchants to upgrade.

However, since we're trying to change the approach in between, meaning we already have stores that utilize the client-side feature flag, we need to support both approaches for one version (`6.6.0`). Once `6.7.0` is released and _all_ stores will use the deferred intent UPE, this complexity will be reduced as everyone will rely on the server-side flag, which we can easily toggle on our side, if needed. 

#### Testing instructions
Check out https://github.com/Automattic/woocommerce-payments-server/pull/3963 in your local server repository.
There are multiple use cases we want to check.

##### Legacy card stores still use legacy card
* Navigate to DevTools, disable [all UPE-related checkboxes](https://github.com/Automattic/woocommerce-payments/assets/22319444/4485110e-da7e-480b-9ebf-d2a77d408b41)
* In account cache, confirm that `'is_deferred_intent_creation_upe_enabled' => false,`. This is because https://github.com/Automattic/woocommerce-payments-server/pull/3963/files#diff-c79fe5913d8e8943eac11e58823d53e42ce3e6234736f0ed0c745fa901b53c17R1844
* Navigate to `Payments -> Settings` and confirm they [look following](https://github.com/Automattic/woocommerce-payments/assets/22319444/7e97ddf0-d725-42f4-8317-79fbed86c9e9)
* Perform a purchase, navigate to payment details on Stripe, and confirm `gateway_type` is `legacy_card`.
* Enable UPE in `Payments -> Settings` by clicking the button from the screenshot linked above
* Perform a purchase, navigate to payment details on Stripe, and confirm `gateway_type` is `split_upe_with_deferred_intent_creation`.
* Navigate to _Payments -> Settings_ and confirm your settings look [like this](https://github.com/Automattic/woocommerce-payments/assets/22319444/35145c7c-9a9f-4602-b2d0-ebb20e183599), meaning without any banners/notices/warnings
* Click kebab menu -> `Disable` and confirm your store is set back to legacy card

##### Legacy UPE stores use deferred UPE
* Navigate to DevTools and set up UPE checkboxes [to look following](https://github.com/Automattic/woocommerce-payments/assets/22319444/1da20ad9-9802-455f-8b84-a8c5fef19c3d)
* In your local server code, change https://github.com/Automattic/woocommerce-payments-server/pull/3963/files#diff-c79fe5913d8e8943eac11e58823d53e42ce3e6234736f0ed0c745fa901b53c17R1844 to compare against `6.5.0` so that your local is treated as eligible, for testing purposes
* Refresh the cache and confirm that `'is_deferred_intent_creation_upe_enabled' => true,`
* Navigate to `Payments -> Settings` and confirm your settings look [like this](https://github.com/Automattic/woocommerce-payments/assets/22319444/7177766e-f2c4-4736-aa27-dd4978433917)
* Confirm that in the kebab menu, there's no `Disable` option
* Run `wp wcpay disable_dupe acct_{YOUR_ACCOUNT_ID}`. With this action, we're testing the process of disabling dUPE for merchants if needed. Refresh the cache afterward.
* Navigate to payments settings and confirm [you're seeing the notice](https://github.com/Automattic/woocommerce-payments/assets/22319444/f9870a97-49f8-4e42-8b56-9d4e35c4468d). (text to be updated, I'm currently aligning the text with Design)
* Run `wp wcpay enable_dupe acct_{YOUR_ACCOUNT_ID}` and refresh the cache
* Confirm that the notice in _Payments -> Settings_ is gone now.
* Perform a payment and confirm that `gateway_type` in Stripe Payment's metadata is `split_upe_with_deferred_intent_creation`.

##### With disabled SEPA payments
* Configure your Stripe account so that SEPA is not active.
* Make sure to have split UPE feature flag enabled in DevTools and confirm that `is_deferred_intent_creation_upe_enabled` is `true` in account cache. 
* Perform a purchase and in Stripe's Payment details, confirm that `gateway_type` is `split_upe_with_deferred_intent_creation`.

##### With enabled SEPA payments
* Configure your Stripe account so that SEPA active.
* Make sure to have split UPE feature flag enabled in DevTools and confirm that `is_deferred_intent_creation_upe_enabled` is `true` in account cache. 
* Perform a purchase and in Stripe's Payment details, confirm that `gateway_type` is `split_upe_with_deferred_intent_creation`.

##### Test any other edge-case that might come to your mind

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `7306` or branch name `feat/handle-server-side-dupe-flag` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=feat/handle-server-side-dupe-flag) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 378641775d6dfab638d844ccc02fa687d5a65526
- Build time: 2023-10-03 09:33:49 UTC

_Note: the build is updated when a new commit is pushed to this PR._

File: client/payment-methods/index.js [hunk 1]
```diff
@@ -287,7 +287,7 @@ const PaymentMethods = () => {
 						>
 							{ interpolateComponents( {
 								mixedString: __(
-									'The new WooPayments checkout experience will become the default on October 11, 2023.' +
+									'You were rolled back from dUPE to legacy UPE. Stay tuned on next actions.' +
```
Comment (timur27) [inline client/payment-methods/index.js]: The wording is currently being aligned with the Design Team. paJDYF-9xf-p2#comment-20974 I'll update the text once confirmed.

File: client/payment-methods/index.js [hunk 2]
```diff
@@ -285,19 +284,8 @@ const PaymentMethods = () => {
...
-								),
-								components: {
-									learnMoreLink: (
-										// eslint-disable-next-line max-len
-										<ExternalLink href="https://woocommerce.com/document/woopayments/payment-methods/additional-payment-methods/#popular-payment-methods" />
-									),
-								},
-							} ) }
+							{ 'You have been switched from the new checkout to your previous checkout experience.' +
+								+'We will keep you posted on the new checkout availability.' }
```
Comment (mdmoore) [inline client/payment-methods/index.js]: I understand this text will be updated soon, but just pointing out the double `+` operator here.

**Review by mdmoore — APPROVED** [review]
Really nice work in reducing the complexity of this task! Everything is testing exactly as described. I noticed one really small thing with the notice after disabling dUPE and noted it for visibility. 

Approving https://github.com/Automattic/woocommerce-payments-server/pull/3963 in tandem with this PR.

### Legacy card stores still use legacy card
✅ In account cache, confirm that `'is_deferred_intent_creation_upe_enabled' => false,`. This is because https://github.com/Automattic/woocommerce-payments-server/pull/3963/files#diff-c79fe5913d8e8943eac11e58823d53e42ce3e6234736f0ed0c745fa901b53c17R1844
✅ Navigate to `Payments -> Settings` and confirm they [look following](https://github.com/Automattic/woocommerce-payments/assets/22319444/7e97ddf0-d725-42f4-8317-79fbed86c9e9)
✅ Perform a purchase, navigate to payment details on Stripe, and confirm `gateway_type` is `legacy_card`.
✅ Enable UPE in `Payments -> Settings` by clicking the button from the screenshot linked above
✅ Perform a purchase, navigate to payment details on Stripe, and confirm `gateway_type` is `split_upe_with_deferred_intent_creation`.
✅ Navigate to _Payments -> Settings_ and confirm your settings look [like this](https://github.com/Automattic/woocommerce-payments/assets/22319444/35145c7c-9a9f-4602-b2d0-ebb20e183599), meaning without any banners/notices/warnings
✅ Click kebab menu -> `Disable` and confirm your store is set back to legacy card

### Legacy UPE stores use deferred UPE
✅ Refresh the cache and confirm that `'is_deferred_intent_creation_upe_enabled' => true,`
✅ Navigate to `Payments -> Settings` and confirm your settings look [like this](https://github.com/Automattic/woocommerce-payments/assets/22319444/7177766e-f2c4-4736-aa27-dd4978433917)
✅ Confirm that in the kebab menu, there's no `Disable` option
✅ Run `wp wcpay disable_dupe acct_{YOUR_ACCOUNT_ID}`. With this action, we're testing the process of disabling dUPE for merchants if needed. Refresh the cache afterward.

🟡 Navigate to payments settings and confirm [you're seeing the notice](https://github.com/Automattic/woocommerce-payments/assets/22319444/f9870a97-49f8-4e42-8b56-9d4e35c4468d). (text to be updated, I'm currently aligning the text with Design)
I understand that this text will be updated. Just highlighting the current text - "You have been switched from the new checkout to your previous checkout experience.NaN"

✅ Run `wp wcpay enable_dupe acct_{YOUR_ACCOUNT_ID}` and refresh the cache
✅ Confirm that the notice in _Payments -> Settings_ is gone now.
✅ Perform a payment and confirm that `gateway_type` in Stripe Payment's metadata is `split_upe_with_deferred_intent_creation`.

### Split UPE stores continue to use deferred UPE
##### With disabled SEPA payments
✅ Configure your Stripe account so that SEPA is not active.
✅ Make sure to have split UPE feature flag enabled in DevTools and confirm that `is_deferred_intent_creation_upe_enabled` is `true` in account cache. 
✅ Perform a purchase and in Stripe's Payment details, confirm that `gateway_type` is `split_upe_with_deferred_intent_creation`.

### With enabled SEPA payments
✅ Configure your Stripe account so that SEPA active.
✅ Make sure to have split UPE feature flag enabled in DevTools and confirm that `is_deferred_intent_creation_upe_enabled` is `true` in account cache. 
✅ Perform a purchase and in Stripe's Payment details, confirm that `gateway_type` is `split_upe_with_deferred_intent_creation`.

File: client/payment-methods/index.js (hunk 2 above)
Comment (timur27) [reply client/payment-methods/index.js]: Text updated in 10b320c41efc5601a873557c64bbbcc88db43ac0 and reflects the latest decisions.
